src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -814,27 +815,50 @@ DagState uniqifyState(ImmutableArray<StateForCase> cases, ImmutableDictionary<Bo
                             break;
                         case BoundDagTest d:
                             bool foundExplicitNullTest = false;
-                            SplitCases(
-                                state.Cases, state.RemainingValues, d,
-                                out ImmutableArray<StateForCase> whenTrueDecisions,
-                                out ImmutableArray<StateForCase> whenFalseDecisions,
-                                out ImmutableDictionary<BoundDagTemp, IValueSet> whenTrueValues,
-                                out ImmutableDictionary<BoundDagTemp, IValueSet> whenFalseValues,
-                                ref foundExplicitNullTest);
-                            state.TrueBranch = uniqifyState(whenTrueDecisions, whenTrueValues);
-                            state.FalseBranch = uniqifyState(whenFalseDecisions, whenFalseValues);
+                            bool foundLengthEvaluation = false;
+                            if (d.Input.Source is { Kind: BoundKind.DagIndexerEvaluation } s0 &&
+                                pendingStates.TryGetValue(s0.Input, out DagState? pendingState))
+                            {
+                                splitCases(currentState: state, processingState: pendingState);
+                                pendingStates.Remove(s0.Input);
+                            }
+
+                            splitCases(state, state);
+
+                            void splitCases(DagState currentState, DagState processingState)
+                            {
+                                SplitCases(
+                                    currentState: currentState, processingState: processingState,
+                                    out ImmutableArray<StateForCase> whenTrueDecisions,
+                                    out ImmutableArray<StateForCase> whenFalseDecisions,
+                                    out ImmutableDictionary<BoundDagTemp, IValueSet> whenTrueValues,
+                                    out ImmutableDictionary<BoundDagTemp, IValueSet> whenFalseValues,
+                                    ref foundExplicitNullTest,
+                                    ref foundLengthEvaluation);
+                                processingState.TrueBranch = uniqifyState(whenTrueDecisions, whenTrueValues);
+                                processingState.FalseBranch = uniqifyState(whenFalseDecisions, whenFalseValues);
```

Comment:
:memo: branches are already set for pending states, we may be able to use these to avoid unnecessary recomputation.